### PR TITLE
Dropping a db fails consistely across adapters.

### PR DESCRIPTION
**Pull request by itsWill**
### Summary

Resolves issue: #32914. When dropping a DB on sqlite3 and on postgresql
if the database (say 'development' and 'test') did not exist the error messages
where inconsistent. On sqlite it would say something like:
```
$ rake db:drop
Dropped database 'db/development.sqlite3'
Database 'db/test.sqlite3' does not exist
```

This was also occuring for postgresql as well. This patch makes the
required fixes to make this experience consistent across adapdters.
After this patch dropping a non existent database should look like:

```
$ rake db:drop
Database 'db/test.sqlite3' does not exist
Database 'db/development.sqlite3' does not exist
```

Across all three: sqlite3, postgres, and mysql adapters.

The issue with `sqlite3` was that when checking if the operation was running on a protected environment we would end up making a connection and in `sqlite3` this causes the database to be created. 

The issue with postgres is that we were rescuing the `ActiveRecord::NoDatabaseError` which causes the incorrect error message to be displayed. 

We fix that by adding a new `db_exists?` method that checks wether a database exists, without making the connection. If the database does not exist we also don't need to do the check protected environment method and we can skip it that check as well. 

### Other Information

There are a couple of things I don't like about this PR. It introduces a bit of complexity and the issue isn't particularly severe, yes the experience is inconsistent but only in the aspect of the messaging the behaviour is still ultimately a correct no-op. 

There are some easter goodies (🐰 🥚 ) that can be extracted out into separate PRs.

### CODE REVIEW

File: activerecord/lib/active_record/tasks/database_tasks.rb
Comment (itsWill) [inline activerecord/lib/active_record/tasks/database_tasks.rb]: This and the `env` method are part of the easter goodies 🐰 🥚 goodies in this PR imo. 

The api for this class allows clients to do something like: 

``` ruby
root = ActiveRecord::Tasks::DatabaseTasks.root
env = ActiveRecord::Tasks::DatabaseTasks.env
puts "Hello from #{root} in #{env}"
```

However this would fail if the `Rails.root` or `Rails.env` are not defined.

File: activerecord/lib/active_record/tasks/database_tasks.rb
Comment (itsWill) [inline activerecord/lib/active_record/tasks/database_tasks.rb]: This method takes care of the case where `Rails.env` is not defined and in fact we use it in a couple of different places e.g. 

* https://github.com/rails/rails/blob/7bc343b3a9ef7581e665a9f70f00d494f7a3b62c/activerecord/lib/active_record/database_configurations/database_config.rb#L33
* https://github.com/rails/rails/blob/7bc343b3a9ef7581e665a9f70f00d494f7a3b62c/activerecord/lib/active_record/migration.rb#L1131

It might make sense to move this out into a `ActiveRecord::Base.current_environment` method and have that be the canonical way of fetching the environment.

**Comment by guilleiguaran** [comment]
@itsWill thanks for working on this!!

Since this is a fix for an existing issue can you extract the easter goodies to separate issue so we can keep this PR only in the scope of the fix? I'm going to merge this after that

**Comment by guilleiguaran** [comment]
I think the current implementation of this might break other adapters that aren't shipped by default with Rails (e.g oracle and sqlserver) and that won't have `db_exists?` initially after of merging this

**Comment by eileencodes** [comment]
> I think the current implementation of this might break other adapters that aren't shipped by default with Rails (e.g oracle and sqlserver) and that won't have db_exists? initially after of merging this

Yea this will and we use a custom database adapter at GitHub that is neither of those. We at least need to have a definition in abstract_adapter that raises a NotImplementedError.

File: activerecord/lib/active_record/tasks/mysql_database_tasks.rb
Comment (eileencodes) [inline activerecord/lib/active_record/tasks/mysql_database_tasks.rb]: I think `database_exists?` is a better method name.

**Review by eileencodes — CHANGES_REQUESTED** [review]
I'm with @guilleiguaran - can you move the goodies into a separate PR so it's easier to review? And change `db_exists?` to `database_exists?`

**Comment by itsWill** [comment]
👋 @guilleiguaran and @eileencodes, thanks for having a look. 

Re: the goodies

The goodies are not completely orthogonal to this PR but they make it signficantly easier to write tests.  Having said that I've removed this one: https://github.com/rails/rails/pull/36025#discussion_r276824531 and stub the root in tests. 

Re: the adapter 

I suspect we're talking of slightly different things here, the adapter is not a subclass of the [AbstractAdapter](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb) and is not part of the `ConnectionAdapters`.

The adapter is one of the task adapters [here](https://github.com/rails/rails/tree/master/activerecord/lib/active_record/tasks). 

Having said that, if someone wrote their own custom task adapter then yes this is a breaking change. I would encourage us to still introduce the `database_exists?` method since it's useful in solving other issues like: https://github.com/rails/rails/issues/36383

**Comment by guilleiguaran** [comment]
> Re: the adapter
> 
> I suspect we're talking of slightly different things here, the adapter is not a subclass of the AbstractAdapter and is not part of the ConnectionAdapters.

Right! But our suggestion is to implement `database_exists?` on database adapter level or in AR::Base.

File: activerecord/lib/active_record/tasks/database_tasks.rb
Comment (itsWill) [reply activerecord/lib/active_record/tasks/database_tasks.rb]: EDIT: proposed another PR for this fix: https://github.com/rails/rails/pull/36435

File: activerecord/test/cases/adapters/mysql2/mysql2_adapter_test.rb
Comment (itsWill) [inline activerecord/test/cases/adapters/mysql2/mysql2_adapter_test.rb]: This makes me a little 😢, when first attempting to connect to a DB the MySQL will raise a unauthorized user error even if the database does not exist: `Mysql2::Error::ConnectionError: Access denied for user 'rails'@'%' to database 'non_extant_database'`. 

The options I saw was to grant the `rails` user permission to the non extant database and then the test works as expected. However instead of managing permissions from our tests I opted for stubbing, it's not ideal but seems to be the best option, I'm open to ideas!

**Comment by itsWill** [comment]
I've moved the `database_exists?` method to the  ConnectionAdapter level, this is good for some 👀

**Comment by itsWill** [comment]
This PR is causing a couple of test failures and has been through a couple of different approaches, I've extracted this PR out: https://github.com/rails/rails/pull/36471 which adds the `database_exists?` to the adapters. This should make it easier to review, and unblock solving other issues that could use this method.  If that gets merged I'll follow up and tackle this issue again.

File: activerecord/test/cases/adapters/mysql2/mysql2_adapter_test.rb
Comment (kamipo) [reply activerecord/test/cases/adapters/mysql2/mysql2_adapter_test.rb]: Probably we can use `inexistent_activerecord_unittest` database for that test.

**Pull request closed without merging**
